pp_netlib/network.py

```diff
@@ -39,82 +41,137 @@ def __init__(self, ref_list, query_list, assignments, outdir, model_type, use_gp
+                self.graph.edge_properties["weight"] = eweight
             else:
-                edges = list(network_data_df.itertuples(index=False, name = None))
-                if previous_network is not None:
-                    for (src, dest) in zip(extra_sources, extra_targets):
-                        edges.append((src, dest))
-        
-            self.graph = gt.Graph(directed = False)
-            self.graph.add_vertex(len(vertex_labels))
+                self.graph.add_edge_list(network_data.values) ## add edges
+
+        ##########################
+        #### SPARSE MAT INPUT ####
+        ##########################
+        elif isinstance(network_data, scipy.sparse.coo_matrix):
+            if not use_gpu:
+                graph_data_df = pd.DataFrame()
+            else:
+                graph_data_df = cudf.DataFrame()
+            graph_data_df["source"] = network_data.row
+            graph_data_df["destination"] =  network_data.col
+            graph_data_df["weights"] = network_data.data
+
+            self.graph.add_vertex(len(graph_data_df)) ## add vertices
+            eweight = self.graph.new_ep("float")
+            self.graph.add_edge_list(list(map(tuple, graph_data_df.values)), eprops = [eweight]) ## add weighted edges
+            self.graph.edge_properties["weight"] = eweight
+
+        ########################
+        ####   LIST INPUT   ####
+        ########################
+        elif isinstance(network_data, list):
+            self.graph.add_vertex(len(network_data)) ## add vertices
+
             if weights is not None:
+                weighted_edges = []
+                for edge in network_data:
+                    weighted_edges.append(edge + (weights[network_data.index(edge)],))
+                
                 eweight = self.graph.new_ep("float")
-                self.graph.add_edge_list(edges, eprops = [eweight])
+                self.graph.add_edge_list(weighted_edges, eprops = [eweight]) ## add weighted edges
                 self.graph.edge_properties["weight"] = eweight
-            else:
-                self.graph.add_edge_list(edges)
             
-        else:
-            # benchmarking concurs with https://stackoverflow.com/questions/55922162/recommended-cudf-dataframe-construction
-            if len(network_data_df) > 1:
-                edge_array = cp.array(network_data_df, dtype = np.int32)
-                edge_gpu_matrix = cuda.to_device(edge_array)
-                gpu_graph_df = cudf.DataFrame(edge_gpu_matrix, columns = ["source","destination"])
             else:
-                # Cannot generate an array when one edge
-                gpu_graph_df = cudf.DataFrame(columns = ["source","destination"])
-                gpu_graph_df["source"] = [network_data_df[0][0]]
-                gpu_graph_df["destination"] = [network_data_df[0][1]]
+                self.graph.add_edge_list(network_data) ## add edges
```

Comment:
I would suggest adding a section after this that adds the names as a vertex property to the graph:
```
v_name_prop = self.graph.new_vertex_property("string")
self.graph.vertex_properties["id"] = vprop
```

Reply:
added
```
v_name_prop = self.graph.new_vp("string")
self.graph.vertex_properties["id"] = v_name_prop
for i in range(len([v for v in self.graph.vertices()])):
    v_name_prop[self.graph.vertex(i)] = vertex_labels[i]
```